service/worker/processor.go

```diff
@@ -176,44 +260,62 @@ func (p *replicationTaskProcessor) worker(workerWG *sync.WaitGroup) {
 				p.logger.Info("Worker for replication task processor shutting down.")
 				return // channel closed
 			}
+			p.processWithRetry(msg, workerID)
+		case <-p.consumer.Closed():
+			p.logger.Info("Consumer closed. Processor shutting down.")
+			return
+		}
+	}
+}
 
-			var err error
-		ProcessRetryLoop:
-			for {
-				select {
-				case <-p.shutdownCh:
-					return
-				default:
-					op := func() error {
-						return p.process(msg)
-					}
-					err = backoff.Retry(op, replicationTaskRetryPolicy, p.isTransientRetryableError)
-					if err != nil && p.isTransientRetryableError(err) {
-						// Keep on retrying transient errors for ever
-						continue ProcessRetryLoop
-					}
+func (p *replicationTaskProcessor) processWithRetry(msg kafka.Message, workerID int) {
+	var err error
+	remainingRetryCount := retryCountInfinity
+
+ProcessRetryLoop:
+	for {
+		select {
+		case <-p.shutdownCh:
+			return
+		default:
+			op := func() error {
+				remainingRetryCount--
+				if remainingRetryCount <= 0 {
+					return errMaxAttemptReached
 				}
 
-				break ProcessRetryLoop
+				errMsg := p.process(msg)
+				if errMsg != nil && p.isTransientRetryableError(errMsg) {
+					// Keep on retrying transient errors for ever
+					p.updateWorkerStatus(workerID, workerStatusPendingRetry)
+					remainingRetryCount = p.getRemainingRetryCount(remainingRetryCount)
+				}
+				return errMsg
 			}
 
-			if err == nil {
-				// Successfully processed replication task.  Ack message to move the cursor forward.
-				msg.Ack()
-			} else {
-				// Task still failed after all retries.  This is most probably due to a bug in replication code.
-				// Nack the task to move it to DLQ to not block replication for other workflow executions.
-				p.logger.WithFields(bark.Fields{
-					logging.TagErr:          err,
-					logging.TagPartitionKey: msg.Partition(),
-					logging.TagOffset:       msg.Offset(),
-				}).Error("Error processing replication task.")
-				msg.Nack()
+			err = backoff.Retry(op, replicationTaskRetryPolicy, p.isTransientRetryableError)
+			if err != nil && p.isTransientRetryableError(err) {
+				// Keep on retrying transient errors for ever
+				continue ProcessRetryLoop
 			}
-		case <-p.consumer.Closed():
-			p.logger.Info("Consumer closed. Processor shutting down.")
-			return
 		}
+
+		p.updateWorkerStatus(workerID, workerStatusRunning)
```

Comment:
I think it is much safer to do this as a defer to make sure all code paths which exit set the status back to running.

Reply:
sure